0015-estimator-interface.md

```diff
@@ -0,0 +1,500 @@
+(This was not always the case via circuit indices, but the history will not be discussed here.)
+It is common, if not typical, for a user to want to estimate many observables corresponding to a single circuit. 
+Likewise, it is common, if not typical, for a user to want to supply multiple parameter value sets for the same circuit.
+This RFC proposes to fix these issues by changing the `Estimator.run()` signature in the following ways: 
+
+ 1. Take the transpose of the current signature; rather than accepting `circuits`, `parameter_values`, and `observables` as three different (iterable) arguments which are to be zipped together, instead localize the distinct tasks to be run via an iterable of triples `(circuit, parameter_values, observables)`.
+ 2. In combination with 1, extend `parameter_values` and `observables` to be array-valued, that is, add the ability to explicity and conveniently specify multiple parameter value sets and observables for a single circuit.
+
+## Motivation <a name="motivation"></a>
+
+Here is a summary of pain-points with the current `Estimator.run()` interface:
+
+1. _Ergonomics._ It feels unnatural to invoke the estimator as `Estimator.run([circuit, circuit], observables=[obs1, obs2])` because of the redundancy of entering the circuit twice. 
+
+1. _Trust._ Cases like `Estimator.run([circuit, circuit], observables=[obs1, obs2])` make a user question whether the primitive implementation is going to do any optimizations to check if the same circuit appears twice, and if so, whether it will be done by memory location, circuit equality, etc.; the interface itself creates gap in trust.
+
+1. _Clarity._ Without reading the documentation in detail, it's not obvious that the invocation `Estimator.run([circuit], observables=[obs1, obs2])` wouldn't cause `circuit` to be run with both supplied observables. In other words, that zipping _is always what's done to the `run()` arguments_ is common source of user confusion. Conversely, it's not clear that the way to collect all combinations of two observables and two parameter sets is to invoke `Estimator.run([circuit] * 4, observables=[obs1, obs2] * 2), parameter_values=[params1] * 2 + [params2] * 2`.
+
+1. _Performance._ Given that `QuantumCircuit` hashing should be avoided and circuit equality checks can be expensive, we should move away from an interface that necessitates performant primitive implementations to go down this path at all. As one example, qubit-wise commuting observables like `"IZZ"` and `"XZI"` can both be estimated using the same simulation/execution, but only if the estimator understands that they share a base circuit. As a second example, when the circuits need to be serialized before they are simulated/executed (e.g. runtime or pickling-for-multi-processing), it puts the onus on the primitive implementation to detect circuit duplications.
```

Comment:
Your comments about the inadequacies of what the primitives do between calls to `run()` are fair, but out of scope; I'd like to see them addressed in a future work. This can be done without compromising the signature proposed here.

This RFC is strictly about the API of the `run()` method, and so really can't make promises about an implementation's internal behaviour. It does, however, address the caching question within a single call. It seems like this won't have much value to your particular workloads, but it does to others, and in particular it has tremendous value to how we get to implement things in deeper areas of our stack.

Reply:
My comment was less about the run signature more about this below, which was talking about expectation of the change around implementation

> hashing should be avoided and circuit equality checks can be expensive, we should move away from an interface that necessitates performant primitive implementations to go down this path at all

If the primitive is just used for one run then the onus would be on the user not to supply duplicates in the inputs anywhere for this to apply (i.e. no duplicate checks needed). If they do have duplicates in input then its down to implementation as to what they do. Same across runs - but without any equality/check caching, performance across runs is going to take a hit and if one has it for that then you have it anyway. It was less clear to me that changing the interface would have any implementation change in that regard as a result of this RFC. The change would make it theoretically possible but for all practical purposes, unless one wanted to take a performance hit across runs - the Session and the primitives was all about improving iterative performance - it seems less likely in that regard. 

But yes,as you say, the RFC is about the interface not the implementation so it maybe a step towards doing the following, at some point in the future if further RFC(s) cover additional changes to allow this to really make this happen.

> we should move away from an interface that necessitates performant primitive implementations to go down this path (_hashing and circuit equality checks) at all